app/src/main/java/org/mozilla/fenix/gleanplumb/state/MessagingMiddleware.kt

```diff
@@ -0,0 +1,85 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.gleanplumb.state
+
+import mozilla.components.lib.state.Middleware
+import mozilla.components.lib.state.MiddlewareContext
+import org.mozilla.fenix.components.appstate.AppAction
+import org.mozilla.fenix.components.appstate.AppAction.MessagingAction.Initialize
+import org.mozilla.fenix.components.appstate.AppAction.MessagingAction.MessageClicked
+import org.mozilla.fenix.components.appstate.AppAction.MessagingAction.MessageDismissed
+import org.mozilla.fenix.components.appstate.AppAction.MessagingAction.UpdateMessages
+import org.mozilla.fenix.components.appstate.AppAction.MessagingAction.MessageDisplayed
+import org.mozilla.fenix.components.appstate.AppAction.MessagingAction.Evaluate
+import org.mozilla.fenix.components.appstate.AppAction.MessagingAction.UpdateMessageToShow
+import org.mozilla.fenix.components.appstate.AppState
+import org.mozilla.fenix.gleanplumb.Message
+import org.mozilla.fenix.gleanplumb.MessagingStorage
+
+class MessagingMiddleware(private val storage: MessagingStorage) : Middleware<AppState, AppAction> {
+
+    override fun invoke(
+        context: MiddlewareContext<AppState, AppAction>,
+        next: (AppAction) -> Unit,
+        action: AppAction
+    ) {
+        when (action) {
+            is Evaluate -> {
+                val nextMessage = storage.getNextMessage()
+                context.dispatch(UpdateMessageToShow(nextMessage))
```

Comment:
see https://github.com/mozilla-mobile/fenix/pull/24225#discussion_r830402471
Because we are using the same action to consume `mesageToShow` in case the conditions are not met anymore.